File: structures/heap.py
"""Heaps!"""
from math import inf
# from heapq import heappop, heappush

# If you can use the built-in heapq
# class Heap(object):
#     def __init__(self, max_heap=False):
#         self.heap = []
#         self.m = -1 if max_heap else 1

#     def __len__(self):
#         return len(self.heap)

#     def peek(self):
#         return self.m * self.heap[0] if self.heap else None

#     def push(self, value):
#         heappush(self.heap, self.m * value)

#     def pop(self):
#         return self.m * heappop(self.heap)


class Heap(object):
    """Binary Min Heap"""

    def __init__(self, max_heap=False):
        self.heap = []
        self.m = -1 if max_heap else 1

    def __len__(self):
        return len(self.heap)

    @staticmethod
    def _parent(index):
        return (index - 1) // 2

    @staticmethod
    def _left_child(index):
        return (index * 2) + 1

    @staticmethod
    def _right_child(index):
        return (index * 2) + 2

    def _swap(self, left, right):
        self.heap[left], self.heap[right] = self.heap[right], self.heap[left]

    def _sift_up(self):
        if self.heap:
            current = len(self.heap) - 1
            parent = self._parent(current)
            while current and self.heap[current] < self.heap[parent]:
                self._swap(current, parent)
                current = parent
                parent = self._parent(current)

    def _node_at(self, index):
        return self.heap[index] if index < len(self.heap) else inf

    def _sift_down(self, current=0):
        if self.heap and current < len(self.heap):
            left, right = self._left_child(current), self._right_child(current)
            if min([current, left, right], key=self._node_at) is not current:
                flip = min([left, right], key=self._node_at)
                self._swap(current, flip)
                self._sift_down(flip)

    def peek(self):
        """View the top of the heap"""
        return self.m * self.heap[0] if self.heap else None

    def push(self, item):
        """Add a new item to the heap"""
        self.heap.append(self.m * item)
        self._sift_up()

    def pop(self):
        "Remove and return the top item from the heap"
        self._swap(0, len(self.heap) - 1)
        item = self.heap.pop()
        self._sift_down()
        return self.m * item


class MinHeap(Heap):
    pass


class MaxHeap(Heap):
    def __init__(self):
        super().__init__(max_heap=True)

File: structures/heap.js
class Heap {
  constructor({ max } = {}) {
    this.heap = [];
    this.m = max ? -1 : 1;
    this.nodeAt = this.nodeAt.bind(this);
  }

  get length() {
    return this.heap.length;
  }

  peek() {
    return this.heap.length ? this.m * this.heap[0] : undefined;
  }

  swap(left, right) {
    [this.heap[left], this.heap[right]] = [this.heap[right], this.heap[left]];
  }

  children(index) {
    return { left: index * 2 + 1, right: index * 2 + 2 };
  }

  parent(index) {
    return Math.floor((index - 1) / 2);
  }

  nodeAt(index) {
    return index < this.heap.length ? this.heap[index] : Infinity;
  }

  siftDown(current = 0) {
    const { left, right } = this.children(current);
    const [c, l, r] = [current, left, right].map(this.nodeAt);
    if (current < this.heap.length && (c > l || c > r)) {
      const flip = l < r ? left : right;
      this.swap(current, flip);
      this.siftDown(flip);
    }
  }

  siftUp(current = this.heap.length - 1) {
    const parent = this.parent(current);
    const [c, p] = [current, parent].map(this.nodeAt);
    if (current > 0 && c < p) {
      this.swap(current, parent);
      this.siftUp(parent);
    }
  }

  pop() {
    let head;
    if (this.heap.length) {
      this.swap(0, this.heap.length - 1);
      head = this.m * this.heap.pop();
      this.siftDown();
    }
    return head;
  }

  push(value) {
    this.heap.push(this.m * value);
    this.siftUp();
  }
}

module.exports = Heap;
